registry/src/api/grpc.rs

```diff
@@ -12,36 +12,110 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::str::FromStr;
+
 use crate::{
     context::Context,
     proto::buffrs::registry::{
         registry_server::Registry, DownloadRequest, DownloadResponse, PublishRequest,
         PublishResponse, VersionsRequest, VersionsResponse,
     },
+    proto::buffrs::package::{
+        Package, Compressed
+    }, types::PackageVersion
 };
 use async_trait::async_trait;
-use tonic::{Request, Response, Status};
+use buffrs::package::PackageName;
+use semver::Version;
+use tonic::{Request, Response, Status, Code};
+use tonic_types::{ErrorDetails, StatusExt};
+
+/// Package's name characters limit, maybe should make it an option later
+const PACKAGE_NAME_LIMIT: usize = 255;
 
 #[async_trait]
 impl Registry for Context {
     async fn publish(
         &self,
-        _request: Request<PublishRequest>,
+        request: Request<PublishRequest>,
     ) -> Result<Response<PublishResponse>, Status> {
-        todo!()
+        let req: PublishRequest = request.into_inner();
+
+        // this shouldn't happen as package isn't optional
+        let package: Compressed = match req.package {
+            None => { return Err(Status::invalid_argument("package wasn't set")); },
+            Some(_value) => {_value}
+        };
+
+        // this shouldn't happen as metadata isn't optional
+        let metadata = match package.metadata {
+            None => { return Err(Status::invalid_argument("metadata wasn't set")); },
+            Some(_value) => {_value} 
+        };
```

Comment:
Same applies here :)

match statements are generally best when you need to branch a lot, if you just want to unpack something there are easier ways :)